### PR TITLE
reimplement find_map(): use the scipy.spatial.cKDtree implementation

### PR DESCRIPTION
**Pull request by vlukes**

### CODE REVIEW

File: sfepy/discrete/fem/mesh.py [hunk 1]
```diff
@@ -7,50 +7,46 @@
 from .meshio import MeshIO, supported_cell_types
 import six
 from six.moves import range
```
Comment (rc) [inline sfepy/discrete/fem/mesh.py]: This is now unused.

File: sfepy/discrete/fem/mesh.py [hunk 2]
```diff
@@ -7,50 +7,46 @@
...
-
-    xs = xx.T[iis]
-    xd = nm.sqrt(nm.sum(nm.diff(xs, axis=0)**2.0, axis=1))
-
-    ii = nm.where(xd < eps)[0]
-    off1, off2 = ir[iis][ii], ir[iis][ii+1]
-    i1, i2 = iis[ii] - off1, iis[ii+1] - off2
-    dns = nm.where(off1 == off2)[0]
-    if dns.size:
+    x = nm.vstack([x1, x2])
```
Comment (rc) [inline sfepy/discrete/fem/mesh.py]: x is not used...

File: sfepy/discrete/fem/mesh.py [hunk 3]
```diff
@@ -7,50 +7,46 @@
...
+    kdtree = cKDTree(nm.vstack([x1, x2]))
```
Comment (rc) [inline sfepy/discrete/fem/mesh.py]: x should go here.

File: sfepy/discrete/fem/mesh.py [hunk 4]
```diff
@@ -7,50 +7,46 @@
...
-                output('x2: %d %d -> %s %s' % (i1[dn], i2[dn],
-                                               x2[:,i1[dn]], x2[:,i2[dn]]))
+        for dn in dns1:
+            idxs = cmap[dn, :]
+            output('x1: %d %d -> %s %s' % (idxs[0], idxs[1],
+                                           x1[idxs[0], :], x1[idxs[1], :]))
+        for dn in dns2:
+            idxs = cmap[dn, :]
+            output('x2: %d %d -> %s %s' % (idxs[0], idxs[1],
+                                           x1[idxs[0], :], x1[idxs[1], :]))
```
Comment (rc) [inline sfepy/discrete/fem/mesh.py]: Should not this be with x2?

**Comment by rc** [comment]
Thanks, good idea! kdtree.query_pairs() seems like a useful function indeed.

I am not sure whether a global `eps` is good, though. I am aware that in periodic.py it is used as well, but nowadays I find passing this as a function parameter much cleaner.

File: sfepy/discrete/fem/mesh.py (hunk 4 above)
Comment (vlukes) [reply sfepy/discrete/fem/mesh.py]: Yes, it should be.

File: sfepy/discrete/fem/mesh.py (hunk 2 above)
Comment (vlukes) [reply sfepy/discrete/fem/mesh.py]: Removed

File: sfepy/discrete/fem/mesh.py (hunk 1 above)
Comment (vlukes) [reply sfepy/discrete/fem/mesh.py]: Removed

**Comment by vlukes** [comment]
Mesh `eps` should be controlled somehow globally. This is a possible way and I think not so bad. Now, I have no idea how to make it better. Passing it as a function argument is clean, but it is not easy to link the `eps` options to all involved functions.

**Comment by rc** [comment]
Thanks for the corrections. 

It seems that the functions involved already had this argument, or you think there will be other functions?

We may come up with something better later, e.g. a context manager, or make eps an attribute of each mesh object?

**Comment by vlukes** [comment]
The mesh functions have `eps` argument, but passing the mesh `eps` option defined in the problem description file to these function is not so straight forward. But, of course, somehow possible.

Yes, maybe make `eps` a mesh object attribute is the solution.

**Comment by rc** [comment]
OK, let us merge this as it is now.

**Comment by rc** [comment]
OK, let us merge this as it is now. BTW. the new problem configuration option(s) should be documented in http://sfepy.org/doc-devel/users_guide.html#miscellaneous.

**Pull request merged**